dask_cuda/device_host_file.py

```diff
@@ -0,0 +1,104 @@
+from zict import Buffer, File, Func
+from zict.common import ZictBase
+from distributed.protocol import deserialize_bytes, serialize_bytes
+from distributed.worker import weight
+
+from functools import partial
+import os
+
+
+def _is_device_object(obj):
+    """
+    Check if obj is a device object, by checking if it has a
+    __cuda_array_interface__ attributed
+    """
+    return hasattr(obj, "__cuda_array_interface__")
+
+
+def _serialize_if_device(obj):
+    """ Serialize an object if it's a device object """
+    if _is_device_object(obj):
+        return serialize_bytes(obj, on_error="raise")
+    else:
+        return obj
+
+
+def _deserialize_if_device(obj):
+    """ Deserialize an object if it's an instance of bytes """
+    if isinstance(obj, bytes):
+        return deserialize_bytes(obj)
+    else:
+        return obj
+
+
+class DeviceHostFile(ZictBase):
+    """ Manages serialization/deserialization of objects.
+
+    Three LRU cache levels are controlled, for device, host and disk.
+    Each level takes care of serializing objects once its limit has been
+    reached and pass it to the subsequent level. Similarly, each cache
+    may deserialize the object, but storing it back in the appropriate
+    cache, depending on the type of object being deserialized.
+
+    Parameters
+    ----------
+    device_memory_limit: int
+        Number of bytes of CUDA device memory for device LRU cache,
+        spills to host cache once filled.
+    memory_limit: int
+        Number of bytes of host memory for host LRU cache, spills to
+        disk once filled.
+    local_dir: path
+        Path where to store serialized objects on disk
+    """
+
+    def __init__(
+        self, device_memory_limit=None, memory_limit=None, local_dir="dask-worker-space"
+    ):
+        path = os.path.join(local_dir, "storage")
+
+        self.host_func = dict()
+        self.disk_func = Func(
+            partial(serialize_bytes, on_error="raise"), deserialize_bytes, File(path)
+        )
+        self.host_buffer = Buffer(
+            self.host_func, self.disk_func, memory_limit, weight=weight
+        )
+
+        self.device_func = dict()
+        self.device_host_func = Func(
+            _serialize_if_device, _deserialize_if_device, self.host_buffer
+        )
+        self.device_buffer = Buffer(
+            self.device_func, self.device_host_func, device_memory_limit, weight=weight
+        )
+
+        self.device = self.device_buffer.fast
+        self.host = self.host_buffer.fast
+        self.disk = self.host_buffer.slow
```

Comment:
It looks like these are LRU objects.  maybe 

```python
        self.device = self.device_buffer.fast.d
        self.host = self.host_buffer.fast.d
        self.disk = self.host_buffer.slow
```

Reply:
I'm sorry, I thought the LRU is what we wanted give access to the users, no? They can give more information, such as the used and total memory from the LRU, not just what objects are in there.

Reply:
By the way, print `self.host` (or `CUDAWorker.data.host`) should be equivalent of `Worker.data.fast`, where I don't think there was any alias such as `Worker.data.fast = Worker.data.fast.d`. Therefore, I think the way it is right now is closer to the behavior from the original `Worker` class.

Reply:
I think that most users will be somewhat confused by the `LRU` object and find it unfamiliar.  I would just hand them the `dict`.  If they want fancier information they can dive through the full object.

Open to counter-arguments though.  I don't have strong confidence in what will happen.

Users are confused today by `Worker.data.fast` (it's not at all clear by the name that this refers to the in-memory data).  I think that we can improve upon the current `Worker.data` approach rather than copy it.

Reply:
> I think that most users will be somewhat confused by the `LRU` object and find it unfamiliar. I would just hand them the `dict`. If they want fancier information they can dive through the full object.
> 
> Open to counter-arguments though. I don't have strong confidence in what will happen.

I see your point, and I think there are two benefits of keeping it the way it is:

1. Users can read the LRU information;
1. Users can still get the dict by simply adding `.d`.

Getting the LRU if they only have the dict is not possible, therefore, I would be more fond of having another alias for the dict (`device_dict`, and such, for example) itself than not having the LRU accessible. However, I have no strong objections to not exposing the LRU, just think it may be useful for some people.

> Users are confused today by Worker.data.fast (it's not at all clear by the name that this refers to the in-memory data). I think that we can improve upon the current Worker.data approach rather than copy it.

I understand that now (there was no documentation for this). In that sense, I would suggest having the same aliases (`host` and `disk`) there too.

Reply:
> Users can still get the dict by simply adding .d.

I think it's unlikley that the average user will know to add `.d`.  We haven't told them, and, even if we did, they wouldn't read about it.

If they want LRU information (which I think will be the uncommon case) then they can go up to the full `device_buffer` object and navigate through.  So the sophisticated results are still available to the sophisticated results, but the simple results are simple to interpret by a simple user.

> I understand that now (there was no documentation for this). In that sense, I would suggest having the same aliases (host and disk) there too.

Agreed.  (though I would rename `host` to `memory` for most users).  Unfortunately we can't add this to the Buffer class (which is a bit more generic) but we could monkey-patch them in this worker code.

https://github.com/dask/distributed/blob/a61df1f54a7ae38c26d8d40de4e1e944067b8dea/distributed/worker.py#L479-L486

Want to submit a short PR or issue?

Reply:
> If they want LRU information (which I think will be the uncommon case) then they can go up to the full `device_buffer` object and navigate through. So the sophisticated results are still available to the sophisticated results, but the simple results are simple to interpret by a simple user.

Fine by me, I will do the change.

> Agreed. (though I would rename `host` to `memory` for most users). Unfortunately we can't add this to the Buffer class (which is a bit more generic) but we could monkey-patch them in this worker code.

That's what I had mind, more or less the same I did here.

> Want to submit a short PR or issue?

Sure, let me first finish the changes here.